### PR TITLE
Microseconds

### PR DESCRIPTION
**Pull request by dannyroberts**
I'm kind of piloting the overall structure of what I'm going to do with timezones by doing the smaller job of adding μs support to datetimes in forms and cases (going forward, on new submissions).

### CODE REVIEW

File: corehq/ext/jsonobject.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# this is like jsonobject.api.re_datetime,
+# but without the "time" part being optional
+# i.e. I just removed (...)? surrounding the second two lines
+re_loose_datetime = re.compile(
+    r'^(\d{4})\D?(0[1-9]|1[0-2])\D?([12]\d|0[1-9]|3[01])'
+    r'\D?([01]\d|2[0-3])\D?([0-5]\d)\D?([0-5]\d)?\D?(\d{3,6})?'
+    r'([zZ]|([\+-])([01]\d|2[0-3])\D?([0-5]\d)?)?$'
+)
+
+class TransDateTimeMeta(object):
```
Comment (dimagimon) [inline corehq/ext/jsonobject.py]: E302 expected 2 blank lines, found 1

**Comment by dannyroberts** [comment]
Going to put this on staging; partly just PRing to run all the tests

**Comment by dannyroberts** [comment]
Yay! Tests passing. I think now's a good time to start reviewing, and I'll throw it up on staging.

@czue @NoahCarnahan

**Comment by NoahCarnahan** [comment]
When should a class inherit from `USecDocument` instead of `Document`?

File: corehq/apps/callcenter/fixturegenerators.py [hunk 2]
```diff
@@ -13,17 +15,26 @@ def should_sync(domain, last_sync, utcnow=None):
...
-    # check if user has already synced today (in local timezone). Indicators only change daily.
-    last_sync_utc = last_sync.date if last_sync.date.tzinfo else utc.localize(last_sync.date)
-    last_sync_local = timezone.normalize(last_sync_utc.astimezone(timezone))
+    _assert = soft_assert(to=['droberts' + '@' + 'dimagi.com'])
+
+    last_sync_utc = last_sync.date
+
+    if not _assert(last_sync_utc.tzinfo is None,
+                   'last_sync.date should be an offset-naive dt'):
+        last_sync_utc = UserTime(last_sync_utc).server_time().done()
```
Comment (snopoke) [inline corehq/apps/callcenter/fixturegenerators.py]: Is `last_sync.date` server time?

https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/restore.py#L437

Comment (snopoke) [inline corehq/apps/callcenter/fixturegenerators.py]: what's the scenario where this would happen?

**Comment by czue** [comment]
@dannyroberts fyi, i'm hoping to do a pass on this this afternoon.

File: corehq/apps/commtrack/models.py [hunk 3]
```diff
@@ -780,10 +782,10 @@ class Meta:
...
     # we need to discuss what will be eventually so we know what we need
     # to support here
-    requested_on = DateTimeProperty()
-    approved_on = DateTimeProperty()
-    fulfilled_on = DateTimeProperty()
-    received_on = DateTimeProperty()
+    requested_on = USecDateTimeProperty()
+    approved_on = USecDateTimeProperty()
+    fulfilled_on = USecDateTimeProperty()
+    received_on = USecDateTimeProperty()
```
Comment (czue) [inline corehq/apps/commtrack/models.py]: yowsers i totally forgot about requisitions. I wonder if we should kill them. (emailed a few people to discuss)

**Comment by czue** [comment]
:+1: from me generally. not sure what we should be worried about breaking on the QA side

File: corehq/apps/callcenter/fixturegenerators.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/callcenter/fixturegenerators.py]: Well, I think it wouldn't happen, and tracing the code it looks like it never happens, but the code I'm replacing had a provision for it having a tzinfo, so this is my no-risk way of asserting it doesn't.

**Comment by dannyroberts** [comment]
@NoahCarnahan Inheriting from `USecDocument` affects the behavior of "dynamic properties", i.e. properties that are not in the schema. Since form and case data is not defined in the schema, these are affected by that. Since datetimes are serialized to json as strings, if you're given a string and it looks like a datetime but it's not part of a schema, couchdbkit (and jsonobject after it) assumes it is a datetime. Using `USecDocument` changes the definition of "looks like a datetime" by changing the regex and then uses `USecDateTimeProperty` instead of `DateTimeProperty` to govern the semantics of it once it's been determined a string is a serialized date.

**Comment by dannyroberts** [comment]
@czue part of my impetus for https://github.com/dimagi/commcare-hq/pull/6213 was to have an easy function to use instead of all the places that currently do a `strptime('%Y-%m-%dT%H:%M:%SZ')`. Doing that will, I imagine, get rid of the vast majority of places where this could go wrong.

**Comment by NoahCarnahan** [comment]
Ah good to know thanks.

**Comment by czue** [comment]
cool. @dannyroberts I'm signed off from a code perspective. defer to your judgment on QA or whether to pull the trigger.

**Comment by dannyroberts** [comment]
Pulled out some more innocuous changes and put them here: https://github.com/dimagi/commcare-hq/pull/6235 to be merged faster and not create one giant diff. Once that's merged I have a few more changes building up on that that I'll add to this branch

File: corehq/apps/hqadmin/views.py [hunk 4]
```diff
@@ -469,7 +470,7 @@ def noneulized_users(request, template="hqadmin/noneulized_users.html"):
     users = WebUser.view("eula_report/noneulized_users",
         reduce=False,
         include_docs=True,
-        startkey =["WebUser", days_ago.strftime("%Y-%m-%dT%H:%M:%SZ")],
+        startkey =["WebUser", datetime_to_iso_string(days_ago)],
```
Comment (dimagimon) [inline corehq/apps/hqadmin/views.py]: E251 unexpected spaces around keyword / parameter equals

**Pull request closed without merging**